eras/babbage/impl/src/Cardano/Ledger/Babbage/TxBody.hs

```diff
@@ -0,0 +1,805 @@
+certs' :: TxBody era -> StrictSeq (DCert (Crypto era))
+txfee' :: TxBody era -> Coin
+wdrls' :: TxBody era -> Wdrl (Crypto era)
+vldt' :: TxBody era -> ValidityInterval
+update' :: TxBody era -> StrictMaybe (Update era)
+reqSignerHashes' :: TxBody era -> Set (KeyHash 'Witness (Crypto era))
+adHash' :: TxBody era -> StrictMaybe (AuxiliaryDataHash (Crypto era))
+mint' :: TxBody era -> Value (Crypto era)
+scriptIntegrityHash' :: TxBody era -> StrictMaybe (ScriptIntegrityHash (Crypto era))
+inputs' (TxBodyConstr (Memo raw _)) = _inputs raw
+
+txnetworkid' :: TxBody era -> StrictMaybe Network
+
+collateral' (TxBodyConstr (Memo raw _)) = _collateral raw
+
+collateralReturn' (TxBodyConstr (Memo raw _)) = _collateralReturn raw
+
+outputs' (TxBodyConstr (Memo raw _)) = _outputs raw
+
+certs' (TxBodyConstr (Memo raw _)) = _certs raw
+
+wdrls' (TxBodyConstr (Memo raw _)) = _wdrls raw
+
+txfee' (TxBodyConstr (Memo raw _)) = _txfee raw
+
+vldt' (TxBodyConstr (Memo raw _)) = _vldt raw
+
+update' (TxBodyConstr (Memo raw _)) = _update raw
+
+reqSignerHashes' (TxBodyConstr (Memo raw _)) = _reqSignerHashes raw
+
+adHash' (TxBodyConstr (Memo raw _)) = _adHash raw
+
+mint' (TxBodyConstr (Memo raw _)) = _mint raw
+
+scriptIntegrityHash' (TxBodyConstr (Memo raw _)) = _scriptIntegrityHash raw
+
+txnetworkid' (TxBodyConstr (Memo raw _)) = _txnetworkid raw
+
+--------------------------------------------------------------------------------
+-- Serialisation
+--------------------------------------------------------------------------------
+
+instance
+  ( Era era,
+    Compactible (Core.Value era)
+  ) =>
+  ToCBOR (TxOut era)
+  where
+  toCBOR (TxOutCompact addr cv) =
+    encodeListLen 2
+      <> toCBOR addr
+      <> toCBOR cv
+  toCBOR (TxOutCompactDH addr cv dh) =
+    encodeListLen 3
+      <> toCBOR addr
+      <> toCBOR cv
+      <> toCBOR dh
+  toCBOR x =
+    let (addr, cv, dh) = viewCompactTxOut x
+     in encodeListLen 3
+          <> toCBOR addr
+          <> toCBOR cv
+          <> toCBOR dh
+
+instance
+  ( Era era,
+    DecodeNonNegative (Core.Value era),
+    Show (Core.Value era),
+    Compactible (Core.Value era)
+  ) =>
+  FromCBOR (TxOut era)
+  where
+  fromCBOR = do
+    lenOrIndef <- decodeListLenOrIndef
+    case lenOrIndef of
+      Nothing -> do
+        a <- fromCBOR
+        cv <- decodeNonNegative
+        decodeBreakOr >>= \case
+          True -> pure $ TxOutCompact a cv
+          False -> do
+            dh <- fromCBOR
+            decodeBreakOr >>= \case
+              True -> pure $ TxOutCompactDH a cv dh
+              False -> cborError $ DecoderErrorCustom "txout" "Excess terms in txout"
+      Just 2 ->
+        TxOutCompact
+          <$> fromCBOR
+          <*> decodeNonNegative
+      Just 3 ->
+        TxOutCompactDH
+          <$> fromCBOR
+          <*> decodeNonNegative
+          <*> fromCBOR
+      Just _ -> cborError $ DecoderErrorCustom "txout" "wrong number of terms in txout"
```

Comment:
Here is a good example of the problem from my other comment. Using TxOutCompact and TxOutCompactDH completely bypasses the ada only optimizations.

Also this deserialization does not account for decoding TxOut with datums